pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/datadriven"
+	"github.com/stretchr/testify/assert"
+)
+
+// TestDataDriven is a data-driven test framework for the schemafeed.
+// It provides a mechanism to classify and understand how the schemafeed
+// will interpret different schema change operations.
+//
+// - "exec"
+//   Executes the input SQL query.
+//
+// - "create" f=<int>
+//   Creates a schemafeed with the targets specified as the input with the
+//   provided ID.
+//
+// - "pop" f=<int>
+//   Pop all events from the schemafeed with the given ID.
+//
+func TestDataDriven(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	defer log.Scope(t).Close(t)
+
+	testData := testutils.TestDataPath(t, "")
+	datadriven.Walk(t, testData, func(t *testing.T, path string) {
+		ctx := context.Background()
+		ts, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+			// TODO(ajwerner): Otherwise the test gets skipped due to some CCL warning.
+			DisableDefaultTestTenant: true,
+		})
+		tdb := sqlutils.MakeSQLRunner(sqlDB)
+		defer ts.Stopper().Stop(ctx)
+		ctx, cancel := ts.Stopper().WithCancelOnQuiesce(ctx)
+		defer cancel()
+		schemaFeeds := map[int]schemafeed.SchemaFeed{}
+		parseTargets := func(t *testing.T, in string) (targets changefeedbase.Targets) {
+			tables := strings.Fields(in)
+			for _, tab := range tables {
+				var tableID descpb.ID
+				var statementTimeName changefeedbase.StatementTimeName
+				tdb.QueryRow(t, "SELECT $1::regclass::int, $1::regclass::string", tab).Scan(
+					&tableID, &statementTimeName)
+				targets.Add(changefeedbase.Target{
+					Type:              jobspb.ChangefeedTargetSpecification_PRIMARY_FAMILY_ONLY,
+					TableID:           tableID,
+					FamilyName:        "primary",
+					StatementTimeName: statementTimeName,
+				})
+			}
+			return targets
+		}
+		errCh := make(chan error, 1)
+		defer func() {
+			select {
+			case err := <-errCh:
+				t.Fatalf("unexpected error: %v", err)
+			default:
+			}
+		}()
+		datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
+			select {
+			case err := <-errCh:
+				d.Fatalf(t, "unexpected error: %v", err)
+			default:
+			}
+			t.Log(d.Cmd)
+			switch d.Cmd {
+			case "exec":
+				if _, err := sqlDB.Exec(d.Input); err != nil {
+					d.Fatalf(t, "failed to exec: %v", err)
+				}
+				return ""
+			case "create":
+				var i int
+				d.ScanArgs(t, "f", &i)
+				if _, exists := schemaFeeds[i]; exists {
+					d.Fatalf(t, "feed %d already created", i)
+				}
+				cfg := &ts.SQLServer().(*sql.Server).GetExecutorConfig().DistSQLSrv.ServerConfig
+				now := ts.Clock().Now()
+				targets := parseTargets(t, d.Input)
+				f := schemafeed.New(ctx, cfg, schemafeed.TestingAllEventFilter, targets, now, nil, changefeedbase.CanHandle{
+					MultipleColumnFamilies: true,
+					VirtualColumns:         true,
+				})
+				schemaFeeds[i] = f
+
+				go func() {
+					if err := f.Run(ctx); !assert.NoError(t, err) {
```

Comment:
i thought it's not safe to require/assert inside go routines like that...?
tests appear to be stuck, I think?

Reply:
`assert` is safe, `require` is not.